File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(head == nullptr){
            return nullptr;
        }
        
        ListNode* prev = new ListNode(head->val);

        while(head->next != nullptr){
            head = head->next;
            prev = new ListNode(head->val,prev);
        }

        return prev;
    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        sp, fp = head,head
        while fp.next and fp.next.next:
            sp = sp.next
            fp = fp.next.next
        second = sp.next
        sp.next = None

        prev = None
        curr = second
        while curr:
            n = curr.next
            curr.next = prev
            prev = curr
            curr = n
        
        first = head
        second = prev

        while second:
            fn = first.next
            sn = second.next

            first.next = second
            second.next = fn

            first = fn
            second = sn



        return